### ISSUE TITLE
[BUG] Schema-invalid serialized result when multiple licenses

### ISSUE THREAD

**Issue by madpah** [CycloneDX/cyclonedx-python-lib#365]
As of `cyclonedx-python-lib` `4.0.0` there appears to be a serialization error when the provided (valid) Model has more than one license added.

Example:
```
bom = Bom()
...
bom.metadata.licenses = [LicenseChoice(license=License(
    id='Apache-2.0', text=AttachedText(
        content='VGVzdCBjb250ZW50IC0gdGhpcyBpcyBub3QgdGhlIEFwYWNoZSAyLjAgbGljZW5zZSE=', encoding=Encoding.BASE_64
    ), url=XsUri('https://www.apache.org/licenses/LICENSE-2.0.txt')
)), LicenseChoice(license=License(name='OSI_APACHE'))]
```
produces
```
...
<ns0:licenses>
    <ns0:license>
        <ns0:id>Apache-2.0</ns0:id>
        <ns0:text content-type="text/plain" encoding="base64">
            VGVzdCBjb250ZW50IC0gdGhpcyBpcyBub3QgdGhlIEFwYWNoZSAyLjAgbGljZW5zZSE=
        </ns0:text>
        <ns0:url>https://www.apache.org/licenses/LICENSE-2.0.txt</ns0:url>
    </ns0:license>
</ns0:licenses>
<ns0:licenses>
    <ns0:license>
        <ns0:name>OSI_APACHE</ns0:name>
    </ns0:license>
</ns0:licenses>
...
```

Which is invalid as per the CycloneDX schema.

this CDX schema discrepancy was fixed via https://github.com/CycloneDX/specification/issues/204

-----

important for the fix: 
https://github.com/CycloneDX/cyclonedx-python-lib/issues/365#issuecomment-1748707040

**Comment by madpah** [comment]
Confirmed this impacts JSON and XML serialization.

**Comment by madpah** [comment]
After some [internal] discussions with CycloneDX folks, we agree that the current Pythonic Model (use of `LicenseChoice`) is likely a poor representation of the CDX Specification.

Definitions such as:
```
def licenses(self) -> "SortedSet[LicenseChoice]":
```

are incorrect as it allows users of the model to supply a set with more than 1 License Expression - which is invalid according to CycloneDX Spec v1.4.

Syntactically, the above definition should more accurately be written as:
```
def licenses(self) -> "Optional[Union[LicenseExpression, Iterable[License]]]":
```

FYI @jkowalleck

**Comment by madpah** [comment]
After further investigation, the challenge associated with serialization for `licenses` is also challenging because the structure/schema is different between JSON and XML under v1.4:

**JSON**
`licenses` is an Array of either `expression` or `license` - i.e. 0 or more `expression` or 0 or more `license` [see here](https://cyclonedx.org/docs/1.4/json/#components_items_licenses_items_license)

**XML**
`licenses` is an complex type defined as either 0 or more `license` OR 0 or 1 `expression` [see here](https://cyclonedx.org/docs/1.4/xml/#type_licenseChoiceType)

**Comment by jkowalleck** [comment]
re: https://github.com/CycloneDX/cyclonedx-python-lib/issues/365#issuecomment-1493806791
this is bad news in terms of a **de**serialization-conflict, because we cannot control input data.
WE can control outputdata via our own serializers - so this path has no problems.

anyway, @madpah , could you open an issue/discussion in https://github.com/CycloneDX/specification/ ?

**Comment by gruebel** [comment]
hey, @jkowalleck and @madpah any update on this? I wanted to upgrade our project to use the latest version ,but this blocking me a bit. Any insights of potential fixes or recommendation on what to do with the multiple license case?

**Comment by jkowalleck** [comment]
I am looking into the proposed fixed.
trying to find a solution.

**Comment by jkowalleck** [comment]
@madpah impossible to create a solution with the exising solution for serialization.
The library is just not capable to model these things. 

we need custom (de)serualization here, as me had previously.
i am not certain if this can be hooked into the currently used seralization-framework at all. 
please help.

**Comment by jkowalleck** [comment]
bugfixing this might break things downstream.
so lets put this into next major version

**Comment by jkowalleck** [comment]
with https://github.com/madpah/serializable/releases/tag/v0.13.0 
i should be able solve the issue soon.

**Comment by jkowalleck** [comment]
implementation details:
- models must be backwards-compatible: accept a mixture of exptession and others
- deserializer must be backwards-compatible: accept a mixture of exptession and others
- serializer: can print just the one appropriate value ...

**Comment by jkowalleck** [comment]
- tests snapshots ready https://github.com/CycloneDX/cyclonedx-python-lib/issues/437
- serialization lib ready  https://github.com/madpah/serializable/releases/tag/v0.14.1
- test data ready https://github.com/CycloneDX/cyclonedx-python-lib/pull/454

----


will tackle this issue soon.

**Comment by jkowalleck** [comment]
solution is implemented and will be part up upcoming major version.
regression test and schema validator in place, to assure a complete fix. :balloon:

**Comment by jkowalleck** [comment]
for everybody willing to test the fix: see [5.0.0-rc.1](https://github.com/CycloneDX/cyclonedx-python-lib/releases/tag/v5.0.0-rc.1)  
this is also available vie PyPI

**Comment by jkowalleck** [comment]
fix available as off https://github.com/CycloneDX/cyclonedx-python-lib/releases/tag/v5.0.0
